sw/host/opentitantool/src/command/gpio.rs

```diff
@@ -153,7 +163,63 @@ impl CommandDispatch for GpioSet {
         transport.capabilities()?.request(Capability::GPIO).ok()?;
         let gpio_pin = transport.gpio_pin(&self.pin)?;
 
-        gpio_pin.set(self.mode, self.value, self.pull)?;
+        gpio_pin.set(self.mode, self.value, self.pull, self.volts)?;
+        Ok(None)
+    }
+}
+
+#[derive(Debug, StructOpt)]
+/// Reads a GPIO pin.
+pub struct GpioAnalogRead {
+    #[structopt(name = "PIN", help = "The GPIO pin to read")]
+    pub pin: String,
+}
+
+#[derive(serde::Serialize)]
+pub struct GpioAnalogReadResult {
+    pub pin: String,
+    pub volts: f32,
+}
+
+impl CommandDispatch for GpioAnalogRead {
+    fn run(
+        &self,
+        _context: &dyn Any,
+        transport: &TransportWrapper,
+    ) -> Result<Option<Box<dyn Annotate>>> {
+        transport.capabilities()?.request(Capability::GPIO).ok()?;
+        let gpio_pin = transport.gpio_pin(&self.pin)?;
+        let volts = gpio_pin.analog_read()?;
+        Ok(Some(Box::new(GpioAnalogReadResult {
+            pin: self.pin.clone(),
+            volts,
+        })))
+    }
+}
+
+#[derive(Debug, StructOpt)]
+/// Writes an analog voltage to a GPIO pin.
+pub struct GpioAnalogWrite {
+    #[structopt(name = "PIN", help = "The GPIO pin to write")]
+    pub pin: String,
+    #[structopt(
+        name = "VOLTS",
+        parse(try_from_str),
+        help = "The analog value to write to the pin in volts, has effect only in AnalogOutput mode"
+    )]
+    pub volts: f32,
```

Comment:
Consider whether we should use `util::voltage::Voltage` here.  Its a thin wrapper around `f64` with a `FromStr` implementation that accepts `v` in the place of the decimal point, as is common practice in our industry (e.g. `3v3` is the same as 3.3 volts).

Its really just a convenience wrapper for CLI parsing.

Reply:
I was not aware of that wrapper.  It makes sense to use it here.  I have used f32 for voltage throughout the traits, since I thought that single precision was more than enough (it takes some serious lab equipment to measure voltage with more than 7 digits of precision.)

Maybe it even makes sense to use the Voltage type in the traits, but I also want the output JSON to be parseable into a floating point by external test scripts, so I do not want "2v123" to appear as the output of `analog-read`, at least not by default.

For now, I have changed just this command argument.